src/lib/update-lock.ts

```diff
@@ -0,0 +1,107 @@
+import * as Bluebird from 'bluebird';
+import * as lockFileLib from 'lockfile';
+import * as _ from 'lodash';
+import { fs } from 'mz';
+import * as path from 'path';
+import * as Lock from 'rwlock';
+
+import constants = require('./constants');
+import { ENOENT, UpdatesLockedError } from './errors';
+
+type asyncLockFile = typeof lockFileLib & {
+	unlockAsync(path: string): Bluebird<void>;
+	lockAsync(path: string): Bluebird<void>;
+};
+const lockFile = Bluebird.promisifyAll(lockFileLib) as asyncLockFile;
+export type LockCallback = (
+	appId: number,
+	opts: { force: boolean },
+	fn: () => PromiseLike<void>,
+) => Bluebird<void>;
+
+function baseLockPath(appId: number): string {
+	return path.join('/tmp/balena-supervisor/services', appId.toString());
+}
+
+export function lockPath(appId: number, serviceName: string): string {
+	return path.join(baseLockPath(appId), serviceName);
+}
+
+function lockFilesOnHost(appId: number, serviceName: string): string[] {
+	return ['updates.lock', 'resin-updates.lock'].map(filename =>
+		path.join(constants.rootMountPoint, lockPath(appId, serviceName), filename),
+	);
+}
+
+const locksTaken: { [lockName: string]: boolean } = {};
+
+// Try to clean up any existing locks when the program exits
+process.on('exit', () => {
+	const result = [];
+	for (const lockName of _.keys(locksTaken)) {
+		try {
+			result.push(lockFile.unlockSync(lockName));
+		} catch (e) {
+			// Ignore unlocking errors
+		}
+	}
+	return result;
```

Comment:
Why are you returning a result to `process.on('exit')`?  I'm guessing this is a coffeescript leftover

Reply:
I've just removed `result` altogether, no need for it - I guess a decaffeinate oddity.

Reply:
It's to match what coffeescript does with it's automatic returns, but we don't care about returning in this case